lib/pos_block_producer.go

```diff
@@ -0,0 +1,206 @@
+	// TODO
+	return nil, nil
+}
+
+// CreateUnsignedBlock constructs an unsigned, PoS block with Fee-Time ordered transactions. This function should be used
+// during happy path in consensus when a vote QC has been assembled. The block is unsigned, so to indicate its incompleteness,
+// the block is returned as a BlockTemplate.
+func (pbp *PosBlockProducer) CreateUnsignedBlock(latestBlockView *UtxoView, newBlockHeight uint64, view uint64,
+	proposerRandomSeedHash *RandomSeedHash, validatorsVoteQC *QuorumCertificate) (BlockTemplate, error) {
+
+	// Create the block template.
+	block, err := pbp.createBlockTemplate(latestBlockView, newBlockHeight, view, proposerRandomSeedHash)
+	if err != nil {
+		return nil, errors.Wrapf(err, "PosBlockProducer.CreateUnsignedTimeoutBlock: Problem creating block template")
+	}
+
+	// Fill out the validators vote QC field.
+	block.Header.ValidatorsVoteQC = validatorsVoteQC
+	return block, nil
+}
+
+// CreateUnsignedTimeoutBlock constructs an unsigned, PoS block with Fee-Time ordered transactions. This function should be used
+// during a timeout in consensus when a validators timeout aggregate QC has been assembled. The block is unsigned,
+// and so is returned as a BlockTemplate.
+func (pbp *PosBlockProducer) CreateUnsignedTimeoutBlock(latestBlockView *UtxoView, newBlockHeight uint64, view uint64,
+	proposerRandomSeedHash *RandomSeedHash, validatorsTimeoutAggregateQC *TimeoutAggregateQuorumCertificate) (BlockTemplate, error) {
+
+	// Create the block template.
+	block, err := pbp.createBlockTemplate(latestBlockView, newBlockHeight, view, proposerRandomSeedHash)
+	if err != nil {
+		return nil, errors.Wrapf(err, "PosBlockProducer.CreateUnsignedTimeoutBlock: Problem creating block template")
+	}
+
+	// Fill out the validators timeout aggregate QC field.
+	block.Header.ValidatorsTimeoutAggregateQC = validatorsTimeoutAggregateQC
+	return block, nil
+}
+
+// createBlockTemplate is a helper function used by CreateUnsignedBlock and CreateUnsignedTimeoutBlock. It constructs
+// a partially filled out block with Fee-Time ordered transactions. The returned block is complete except for
+// the qc / aggregateQc fields, and the signature.
+func (pbp *PosBlockProducer) createBlockTemplate(latestBlockView *UtxoView, newBlockHeight uint64, view uint64,
+	proposerRandomSeedHash *RandomSeedHash) (BlockTemplate, error) {
+	// First get the block without the header.
+	block, err := pbp.createBlockWithoutHeader(latestBlockView, newBlockHeight)
+	if err != nil {
+		return nil, errors.Wrapf(err, "PosBlockProducer.CreateBlockTemplate: Problem creating block without header")
+	}
+
+	// Fill out what we can in the block header. This function won't fill out any of the consensus QC fields.
+	block.Header.Version = HeaderVersion2
+	block.Header.PrevBlockHash = latestBlockView.TipHash
+
+	// Compute the merkle root for the block now that all of the transactions have been added.
+	merkleRoot, _, err := ComputeMerkleRoot(block.Txns)
```

Comment:
I would add the extra info you want to add to the merkle root this ComputeMerkleRoot function with a new argument that is allowed to be nil, and delete the new merkle root field you're adding. Also you neglected to set the new merkle root field here looks like, but again I would just delete it.